### DISCUSSION TITLE
An automatic way to reduce times of visiting useless AST nodes

### DISCUSSION THREAD

**Discussion by hyf0** [swc-project/swc#6612] [Ideas]
I think about this while reducing times of visiting useless AST nodes manually.

let’s say we have a macro called `magic_visit_mut` . We could use it like

```rust
struct OnlyCareModuleItem;

magic_visit_mut!(impl VisitMut for OnlyCareModuleItem {
  fn visit_module_item(&mut self, node: &mut ModuleItem) {
		// do something
  }
})
```

and the macro will expand as

```rust
struct OnlyCareModuleItem;

impl VisitMut for OnlyCareModuleItem {
  // --- noop visiting function automatically generated by  `magic_visit_mut`
  fn visit_mut_script(&mut self, node: &mut Script) {}
  // --- End

  fn visit_module_item(&mut self, node: &mut ModuleItem) {
    // do something
  }
}
```

# algorithm

We will start with the node visiting function that is declared in  `magic_visit_mut`. If a node could be visited, then all possible parents and children of it could be visited. After the analysis, we will generate noop visiting function for non-visited nodes.

And, we could add extra hints for each visitor to help with analysis. Such as `#[magic_visit_mut(wont_visit_children)]`

```rust
struct OnlyCareModuleItem;

magic_visit_mut!(impl VisitMut for OnlyCareModuleItem {
  #[magic_visit_mut(wont_visit_children)]
  fn visit_module_item(&mut self, node: &mut ModuleItem) {
		// do something
  }
})
```

the macro will expand as

```rust
struct OnlyCareModule;

impl VisitMut for OnlyCareModuleItem {
  // --- noop visiting function automatically generated by  `magic_visit_mut`
  fn visit_mut_script(&mut self, node: &mut Script) {}

  fn visit_mut_import_decl(&mut self, node: &mut Script) {}

  fn visit_mut_export_decl(&mut self, node: &mut Script) {}

  fn visit_mut_stmts(&mut self, node: &mut Script) {}

  // ....
  // --- End

  fn visit_module_item(&mut self, node: &mut ModuleItem) {
    // do something
  }
}
```

I’m not sure how much the performance could be improved, but I guess it wasn’t bad to raise an issue to discuss it.

**Comment by kdy1** [comment]
I thought `rustc` would optimize it to noop, but seems like it was not the case?

**Reply by hyf0** [reply]
Yeah, it's not the case.
```rust
struct OnlyCareModuleItem;

impl VisitMut for OnlyCareModuleItem {
  fn visit_module_item(&mut self, node: &mut ModuleItem) {
    // do something
  }
}
```
The above code only overwrites the function `visit_module_item`. The rest of functions are untouched, so, such as `visit_mut_script` will have a default implementation like
```rust
  fn visit_module_script(&mut self, node: &mut Script) {
    node.visit_mut_chidlren_wtih(self)
  }
```

Giving a huge AST `Program::Script` and applying `OnlyCareModuleItem` to the AST, we will still go through all nodes of the AST, though we only care about `ModuleItem`.

We could optimize the code manually and make `visit_module_script` a noop function, but this requires us to execute the above algorithm using our brain, which could be buggy.

**Reply by kdy1** [reply]
Doesn't LTO work? There's no `#[inline]`, so LTO is required to optimize it out.

**Reply by hyf0** [reply]
Not sure, I will look into it.

**Reply by kdy1** [reply]
Did it work?

**Reply by hyf0** [reply]
I haven't investigated it yet. But I found a visitor of SWC only takes 20+ms to visit a huge AST parsed from a 20mb+ javascript file, which makes me feel it's unnecessary to do such optimization.

**Reply by kdy1** [reply]
I see. Thank you for sharing!
